### PR TITLE
Users can post stories.

### LINKED ISSUE #4: Users can post stories.

### PR DESCRIPTION
**Pull request by connorlay**
resolves #4 

Users can post stories that have a title and either a url, text, or both. Stories are displayed on the root page in descending order of creation. A user's profile displays all of the stories they have submitted. A user who is not logged in that attempts to post a new story is redirected to the login page with friendly forwarding.

posting a new story
url validation
without a url
attempting to post a new story when not signed in

### CODE REVIEW

File: app/controllers/stories_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  def index
+    @stories = Story.all
+  end
+
+  private
+
+  def story_params
+    params.require(:story).permit(:title, :url, :text)
+                          .merge(user_id: current_user.id)
```
Comment (efatsi) [inline app/controllers/stories_controller.rb]: another option to solving the User relating issue would be to change line 10 above to be:
```ruby
@story = current_user.stores.new(story_params)
```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -17,6 +17,7 @@ def create
 
   def show
     @user = User.find(params[:id])
+    @stories = Story.where(user_id: @user.id)
```
Comment (efatsi) [inline app/controllers/users_controller.rb]: `@stories = @user.stories` would also do the trick here

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -13,4 +13,25 @@ def log_out
     @current_user = nil
   end
 
+  def logged_in?
+    !current_user.nil?
```
Comment (efatsi) [inline app/helpers/sessions_helper.rb]: i try to avoid double negatives as much as I can in programming (and in life) to preserve my sanity. Occasionally one is necessary, but I'm guessing `current_user.present?` would do just fine here.

File: app/controllers/stories_controller.rb (hunk 1 above)
Comment (Fosome) [inline app/controllers/stories_controller.rb]: :+1: preferred

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -13,4 +13,25 @@ def log_out
...
+  def check_logged_in
+    unless logged_in?
+      store_location
+      flash[:warning] = Messaging::Sessions::not_logged_in
+      redirect_to login_path
+    end
+  end
+
+  def redirect_back_or(default)
+    redirect_to(session[:forwarding_url] || default)
```
Comment (efatsi) [inline app/helpers/sessions_helper.rb]: probably want to add a `.presence` on to `session[:forwarding_url]`. If that were to somehow be set to an empty string, `.presence` would return false and the default would be chosen. Maybe an unnecessary precaution but I feel like that weird stuff happens from time to time when dealing with sessions/cookies

File: app/controllers/stories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+class StoriesController < ApplicationController
+
+  before_action :check_logged_in, only: [ :new, :create ]
```
Comment (Fosome) [inline app/controllers/stories_controller.rb]: I like splitting up the conditional and the action:

`before_action :redirect_to_login, unless: :logged_in?, only: [:new, :create]`

File: app/models/story.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class Story < ActiveRecord::Base
+  validates :title,      presence:    true,
+                         length:      { maximum: 255 }
+  validates :url,        format:      { with: URI::regexp(["http", "https"]) },
+                         allow_blank: true
+  validates :text,       length:      { maximum: 10000 }
+  validates :url, :text, any_one:     true
+
+  belongs_to :user
+  default_scope -> { order(created_at: :desc) }
```
Comment (Fosome) [inline app/models/story.rb]: Avoid `default_scope` where possible.  It can cause some headaches when you need to do something different.

File: app/views/shared/_error_messages.html.erb [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <% if object.errors.any? %>
   <div class="panel panel-danger">
     <div class="panel-heading">
-      <%= "There are #{pluralize(object.errors.count, "error")}:" %>
+      <%= "There #{object.errors.count == 1 ? "is" : "are"} #{pluralize(object.errors.count, "error")}:" %>
```
Comment (efatsi) [inline app/views/shared/_error_messages.html.erb]: Everyone knows to use "be" instead of "is" or "are", for simplicity reasons:
```
There **be** 1 error.
There **be** 5 errors.
```

File: app/models/story.rb (hunk 6 above)
Comment (efatsi) [inline app/models/story.rb]: Aha, was wondering why there wasn't a scope called in the controller. Agree with Ryan

File: config/application.rb [hunk 8]
```diff
@@ -25,5 +25,7 @@ class Application < Rails::Application
 
     config.autoload_paths += %W(#{config.root}/lib)
     config.autoload_paths += Dir["#{config.root}/lib/**/"]
+
+    config.autoload_paths += %W["#{config.root}/app/validators/"]
```
Comment (Fosome) [inline config/application.rb]: Just group all these into a single call.

File: db/migrate/20150622211739_create_stories.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+class CreateStories < ActiveRecord::Migration
+  def change
+    create_table :stories do |t|
+      t.string :url
+      t.string :title
```
Comment (Fosome) [inline db/migrate/20150622211739_create_stories.rb]: required?

File: app/views/stories/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<h1>
+  <%= @story.title %><br>
+  <small><%= link_to @story.url, @story.url, target: "_blank" %></small>
```
Comment (efatsi) [inline app/views/stories/show.html.erb]: Should this be wrapped in a conditional based on the presence of the story's url?

File: lib/messaging.rb [hunk 11]
```diff
@@ -39,6 +39,15 @@ def self.create_success(user)
...
 
     DESTROY_SUCCESS = "Successfully logged out."
+
+    NOT_LOGGED_IN   = "You must login to continue."
+  end
+
+  module Stories
+    extend MethodMissingOverride
+
+    CREATE_SUCCESS  = "Story successfully created."
```
Comment (Fosome) [inline lib/messaging.rb]: Why not `i18n`?   http://guides.rubyonrails.org/i18n.html

File: spec/models/story_spec.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+    context "without text or a url" do
+      subject { build :story, url: "   ", text: "   " }
+      it { is_expected.not_to be_valid }
+    end
+
+    context "with text or a url" do
+      subject { build :story, url: "  ", text: "Some awesome text!" }
+      it { is_expected.to be_valid }
+    end
```
Comment (efatsi) [inline spec/models/story_spec.rb]: would recommend two tests here, one for the text (you already have this), and one for just the url

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -6,14 +6,16 @@
     it { is_expected.to validate_presence_of :name }
     it { is_expected.to validate_length_of(:name).is_at_most 50 }
     it { is_expected.to validate_uniqueness_of :name }
+    it { is_expected.to have_db_index :name }
```
Comment (Fosome) [inline spec/models/user_spec.rb]: Interesting, I've never seen this before. :+1:

File: app/controllers/stories_controller.rb (hunk 5 above)
Comment (connorlay) [inline app/controllers/stories_controller.rb]: That looks much better :+1:

File: app/models/story.rb (hunk 6 above)
Comment (connorlay) [inline app/models/story.rb]: I haven't really used Rails scopes before –this was something from a textbook I read. Thanks for the tip!

File: app/views/stories/show.html.erb (hunk 10 above)
Comment (connorlay) [inline app/views/stories/show.html.erb]: Yes it should!

File: app/controllers/password_reset_controller.rb [hunk 14]
```diff
@@ -11,10 +11,10 @@ def create
     if @user = User.find_by(email: params[:password_reset][:email].downcase)
       @user.create_reset_digest
       @user.send_password_reset_email
-      flash[:info] = Messaging::PasswordResets::create_success
+      flash[:info] = t(:email_sent, scope: :password_resets)
```
Comment (h0tl33t) [inline app/controllers/password_reset_controller.rb]: You can also do `t('password_resets.email_sent')`.

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -7,17 +7,17 @@ def create
     user = User.find_by(email: params[:session][:email].downcase)
     if user && user.authenticate(params[:session][:password])
       log_in user
-      flash[:success] = Messaging::Sessions::create_success(user)
-      redirect_to root_path
+      flash[:success] = t(:login, scope: :sessions, name: user.name)
```
Comment (h0tl33t) [inline app/controllers/sessions_controller.rb]: And when you need to pass something along:
```ruby
t('sessions.login', name: user.name)
```

File: app/validators/any_one_validator.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class AnyOneValidator < ActiveModel::Validator
+
+  def validate(record)
+    attributes = options[:attributes]
+    unless attributes.any? { |attribute| record.send(attribute).present? }
```
Comment (h0tl33t) [inline app/validators/any_one_validator.rb]: The positive case might read a bit better:
```
if attributes.none? { |attribute| record.send(attribute).present? }
```

File: app/views/users/show.html.erb [hunk 17]
```diff
@@ -3,4 +3,8 @@
     <%= @user.name %>
     <small><%= @user.email %></small>
   </h1>
+  <% unless @stories.empty? %>
+    <%= render "/shared/stories_panel", heading: "#{@user.name}'s Stories",
+                                       stories: @stories %>
```
Comment (h0tl33t) [inline app/views/users/show.html.erb]: Spacing :tm:

File: config/locales/en.yml [hunk 18]
```diff
@@ -1,23 +1,19 @@
...
+    email_sent:      "Email sent with password reset instructions."
+    email_not_found: "Email address not found."
+    invalid_reset:   "Invalid email/token."
+    password_reset:  "Password reset. You may now log in with your new password."
+    expired_token:   "Password reset has expired."
+  users:
+    welcome:         "Welcome to the Hacker News Clone, %{name}!"
+  sessions:
+    login:           "Logged in as %{name}."
+    invalid_login:   "Icorrect email/password combination."
```
Comment (h0tl33t) [inline config/locales/en.yml]: Spelling :tm:

**Comment by h0tl33t** [comment]
Nice job!  :+1:

File: app/controllers/password_reset_controller.rb (hunk 14 above)
Comment (ltk) [inline app/controllers/password_reset_controller.rb]: managing all of an app's messages can get unwieldy quickly if you lack organization in naming. One option you could consider for the future would be to name your messages more-or-less RESTfully like `password_resets.create.success`, `password_resets.create.failure.email_not_found`

That approach is certainly imperfect (breaks down particularly when you want to share messages across actions), but more scoped organization of any variety has been helpful to me in the past.

File: app/controllers/stories_controller.rb [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+  before_action :redirect_to_login, unless: :logged_in?, only: [ :new, :create ]
+
+  def new
+    @story = Story.new
+  end
+
+  def create
+    @story = current_user.stories.new(story_params)
+    if @story.save
+      flash[:success] = t(:story_created, scope: :stories)
```
Comment (ltk) [inline app/controllers/stories_controller.rb]: `stories.story_created` seems a bit redundant

File: app/controllers/stories_controller.rb [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+      redirect_to story_path(@story)
```
Comment (ltk) [inline app/controllers/stories_controller.rb]: some would tell you that it's always best to redirect to an absolute URL instead of a relative path. although I believe the RFC spec around redirect responses allows both as of last year.

File: app/models/story.rb [hunk 21]
```diff
@@ -0,0 +1,13 @@
+class Story < ActiveRecord::Base
+  validates :title,      presence:    true,
+                         length:      { maximum: 255 }
+  validates :url,        format:      { with: URI::regexp(["http", "https"]) },
+                         allow_blank: true
+  validates :text,       length:      { maximum: 10000 }
```
Comment (ltk) [inline app/models/story.rb]: why a max length validation on this field?

File: app/views/shared/_stories_panel.html.erb [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    <%= heading %>
+  </div>
+  <ul class="list-group">
+    <% stories.each do |story| %>
+      <li class="list-group-item">
+        <%= link_to story_path(story) do %>
+          <b><%= story.title %></b>
+        <% end %>
+        <%= link_to story.url, story.url %>
+        <%= link_to user_path(story.user.id) do %>
```
Comment (ltk) [inline app/views/shared/_stories_panel.html.erb]: I believe you should be able to just pass the user rather than their id.

File: db/seeds.rb [hunk 23]
```diff
@@ -1,7 +1,10 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
-#
-# Examples:
-#
-#   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
-#   Mayor.create(name: 'Emanuel', city: cities.first)
+10.times do |n|
+  user = User.create(name: Faker::Name.name,
```
Comment (ltk) [inline db/seeds.rb]: you could line up this hash if you wanted to.

File: spec/features/story_creations_spec.rb [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+            before { fill_in_form story, :url }
+
+            it "is redirected to the story page" do
+              expect(current_path).to eq story_path(Story.last.id)
+            end
+            it "sees a title and url" do
+              expect(page).to have_content story.url
+              expect(page).to have_content story.title
+            end
+            it "sees a success message" do
```
Comment (ltk) [inline spec/features/story_creations_spec.rb]: There are differences of opinion on whether or not to group multiple feature-spec assertions like these into a single `it` block. What are the pros/cons?

File: spec/features/story_creations_spec.rb [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+              expect(page).to have_content I18n.t(:story_created, scope: :stories)
+            end
+          end
+        end
+
+        context "submitting invalid parameters" do
+          before { fill_in_form invalid_story, :url }
+
+          it "sees an error message" do
+            expect(page).to have_content "error"
```
Comment (ltk) [inline spec/features/story_creations_spec.rb]: `"error"` is a pretty generic word to test against. Can you be more specific?

File: spec/features/story_creations_spec.rb [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+        before { visit "/stories/new" }
+
+        it "is redirected to the log in page" do
+          expect(current_path).to eq login_path
+        end
+        it "sees an error message" do
+          expect(page).to have_content I18n.t(:not_logged_in, scope: :sessions)
+        end
+
+        context "logging in with valid credentials" do
```
Comment (ltk) [inline spec/features/story_creations_spec.rb]: This spec does not seem related to story creation.

File: spec/features/story_creations_spec.rb [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+
+          it "is redirected to the new story page" do
+            expect(current_path).to eq new_story_path
+          end
+        end
+      end
+    end
+  end
+
+  def fill_in_form(story, attribute)
```
Comment (ltk) [inline spec/features/story_creations_spec.rb]: this name is slightly misleading in that it both fills in the form _and_ submits it.

File: spec/features/user_profiles_spec.rb [hunk 28]
```diff
@@ -4,22 +4,32 @@
...
+      it "sees its stories" do
+        stories = Story.where(user_id: user.id)
+        stories.each do |story|
+          expect(page).to have_content story.title
+          expect(page).to have_content story.url
+        end
+      end
+      it "does not see other users' stories" do
+        other_stories = Story.where.not(user_id: user.id)
+        other_stories.each do |story|
```
Comment (ltk) [inline spec/features/user_profiles_spec.rb]: this is brittle. it's easy to imagine someone else coming into this file, not understanding the importance of `create :user_with_stories` way up the file, and removing it, in which case this would pass incorrectly.

I'd do one of two things:
- more explicitly express the intention around `create :user_with_stories` by turning it into a `let!(other_user) { `create :user_with_stories` }`
- or, create a specific Story to test against `let!(other_users_story) { create(:story) } ... expect(page).not_to have_content other_users_story.title`

**Comment by ltk** [comment]
I'd like to see unit tests for you custom validator. Otherwise :+1:

File: app/models/story.rb (hunk 21 above)
Comment (connorlay) [inline app/models/story.rb]: I'm using postgresql,which does not enforce a max length on text by default. (Although I should probably move this validation to the migration instead with ```t.text :text, limit: 10000```)

File: spec/features/story_creations_spec.rb (hunk 24 above)
Comment (connorlay) [inline spec/features/story_creations_spec.rb]: I should probably separate them. If one of the expectations fails the whole statement fails, which makes debugging more difficult.

File: lib/messaging.rb [hunk 29]
```diff
@@ -1,44 +0,0 @@
...
-    def self.create_success(user)
-      "Successfully logged in as #{user.name}"
-    end
-
-    CREATE_FAILURE  = "Incorrect email/password combination."
-
-    DESTROY_SUCCESS = "Successfully logged out."
-  end
-
-end
```
Comment (efatsi) [inline lib/messaging.rb]: RIP `Messaging`. Twas a good custom solution for the problem!

**Comment by efatsi** [comment]
Nice work :+1:

**Comment by Fosome** [comment]
:+1:

File: lib/messaging.rb (hunk 29 above)
Comment (connorlay) [inline lib/messaging.rb]: It was fun to figure out how to decouple messages from my tests, but I figured convention over configuration is the Rails way so I switched to i18n.
